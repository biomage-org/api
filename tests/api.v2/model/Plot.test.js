// @ts-nocheck
// Disabled ts because it doesn't recognize jest mocks
const config = require('../../../src/config');

const { mockSqlClient } = require('../mocks/getMockSqlClient')();
const getObject = require('../../../src/api.v2/helpers/s3/getObject');

jest.mock('../../../src/api.v2/helpers/s3/getObject');
jest.mock('../../../src/sql/sqlClient', () => ({
  get: jest.fn(() => mockSqlClient),
}));

const Plot = require('../../../src/api.v2/model/Plot');
const BasicModel = require('../../../src/api.v2/model/BasicModel');

const mockExperimentId = 'mockExperimentId';
const mockPlotUuid = 'mockPlotUuid';


describe('model/Plot', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('Get plot config works correctly', async () => {
    const mockConfig = { some: 'config' };
    const mockS3DataKey = 'mock/plot/data/key';
    const mockPlotData = [1, 2, 3];

    const mockFindOne = jest.spyOn(BasicModel.prototype, 'findOne')
      .mockImplementationOnce(() => ({
        s3DataKey: mockS3DataKey,
        config: mockConfig,
      }));

    getObject.mockImplementation(() => (JSON.stringify({
      plotData: mockPlotData,
    })));

    const result = await new Plot().getConfig(mockExperimentId, mockPlotUuid);

    expect(mockFindOne).toHaveBeenCalledTimes(1);
    expect(mockFindOne).toHaveBeenCalledWith({ id: mockPlotUuid, experiment_id: mockExperimentId });

    expect(getObject).toHaveBeenCalledTimes(1);
    expect(getObject).toHaveBeenCalledWith({ Bucket: `plots-tables-${config.clusterEnv}`, Key: mockS3DataKey });

    expect(result).toEqual({ config: mockConfig, plotData: mockPlotData });
  });

  it('Does not download plot data from S3 if plot does not have plot data', async () => {
    const mockConfig = { some: 'config' };

    const mockFindOne = jest.spyOn(BasicModel.prototype, 'findOne')
      .mockImplementationOnce(() => ({
        s3DataKey: null,
        config: mockConfig,
      }));


    const result = await new Plot().getConfig(mockExperimentId, mockPlotUuid);

    expect(mockFindOne).toHaveBeenCalledTimes(1);
    expect(mockFindOne).toHaveBeenCalledWith({ id: mockPlotUuid, experiment_id: mockExperimentId });

    expect(getObject).not.toHaveBeenCalled();

    expect(result).toEqual({ config: mockConfig });
  });

  it('Update plot config works correctly', async () => {
    const mockConfig = {
      legend: { enabled: true },
      plotTitle: 'mockPlotTitle',
    };

    const mockUpdate = jest.spyOn(BasicModel.prototype, 'update');

    await new Plot().updateConfig(mockExperimentId, mockPlotUuid, mockConfig);

    expect(mockUpdate).toHaveBeenCalledTimes(1);
    expect(mockUpdate).toHaveBeenCalledWith({ id: mockPlotUuid, experiment_id: mockExperimentId }, { config: mockConfig });
  });
});
