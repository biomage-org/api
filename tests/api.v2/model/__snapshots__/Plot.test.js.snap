// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`model/Plot invalidateAttributesForMatches works correctly: andWhereLike calls 1`] = `
Array [
  Array [
    "id",
    "ViolinMain%",
  ],
]
`;

exports[`model/Plot invalidateAttributesForMatches works correctly: returning calls 1`] = `
Array [
  Array [
    Array [
      "id",
      "config",
    ],
  ],
  Array [
    Array [
      "id",
      "config",
    ],
  ],
  Array [
    Array [
      "id",
      "config",
    ],
  ],
]
`;

exports[`model/Plot invalidateAttributesForMatches works correctly: select calls 1`] = `
Array [
  Array [
    Array [
      "id",
      "config",
    ],
  ],
]
`;

exports[`model/Plot invalidateAttributesForMatches works correctly: update calls 1`] = `
Array [
  Array [
    Object {
      "config": Object {
        "title": Object {
          "anchor": "start",
          "dx": 0,
          "fontSize": 20,
          "text": "Gzma",
        },
      },
    },
  ],
  Array [
    Object {
      "config": Object {
        "title": Object {
          "anchor": "start",
          "dx": 0,
          "fontSize": 20,
          "text": "Lyz2",
        },
      },
    },
  ],
  Array [
    Object {
      "config": Object {
        "title": Object {
          "anchor": "start",
          "dx": 0,
          "fontSize": 20,
          "text": "Lyz1",
        },
      },
    },
  ],
]
`;

exports[`model/Plot invalidateAttributesForMatches works correctly: where calls 1`] = `
Array [
  Array [
    Object {
      "experiment_id": "mockExperimentId",
    },
  ],
  Array [
    Object {
      "experiment_id": "mockExperimentId",
      "id": "ViolinMain",
    },
  ],
  Array [
    Object {
      "experiment_id": "mockExperimentId",
      "id": "ViolinMain-0",
    },
  ],
  Array [
    Object {
      "experiment_id": "mockExperimentId",
      "id": "ViolinMain-1",
    },
  ],
]
`;
